### PR TITLE
feat: Configurable default backend

### LINKED ISSUE #1004: Explicit setting of the default backend
It may be desirable in some cases (e.g. #882) to have one consistent backend for all cases, which would enable the propagation of gradients throughout pyhf model construction, for instance. Backends will hit a wall when trying to take the gradient with respect to something that involves pure numpy.

An example implementation of this could be as simple as adding the argument `set_default=False` to `set_backend`, followed by something like

```
def set_backend(..., set_default=False)
...
backend = tensorlib
if set_default:
        global default_backend
        default_backend = tensorlib
...
tensorlib._setup()
if set_default:
    default_backend._setup()
```

It seems like this would be fine at first glance, since the backend checks are done in the lines before this. I managed to get the `logpdf` of `hepdata_like` with no issues, but I was unsure what would be a more robust test for changing the default backend. Ideally, one would want to run the whole test suite I guess, to make sure this doesn't secretly change some functionality.

Hope this is useful!

### PR DESCRIPTION
**Pull request by kratsg**
# Pull Request Description

Supersedes #1121. Should allow for default backend to be changeable.

Resolves #1004.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Add conditional logic in `set_backend` and `get_backend` based on
a new kwarg `default=False` that will set or get the default backend
instead of the tensorlib backend. Particularly useful for differentiable
model construction, which needs to be propagated through tensors
instead of numpy arrays.
* Migrate relevant code in src/pyhf/__init__.py into a submodule
instead and clean things up more
* Add tests to ensure autodifferentiability and jit with the
changing of default backends
```

### CODE REVIEW

**Comment by kratsg** [comment]
> @kratsg I think we don't need all the `import pyhf` everywhere. I know that you said you're doing this so avoid circular imports, but I think this can be avoided.

we need this to access `pyhf.default_backend`. I cannot `from pyhf import default_backend` anymore fyi.

**Comment by kratsg** [comment]
> Can you add a test that shows that a function that uses both default backend and tensorlib with both set to jax becomes diffable?

Yes, I can definitely add a test. I'm hoping @phinate can provide some suggestions on a way to ensure diffability.

**Comment by phinate** [comment]
> I would try something like
> 
> def func(x): y = default_backend.power(x, 2) z = tensorlib.power(y, 2) return z
> 
> jax.jacrev(func)(default_backend.astensor([1.,2.]))

Yeah, this is a nice way to check there’s no backend foul play on the scale of individual ops. Probably not a planned use case, but I imagine this is also possible to test for the other backends.

To ensure the most general case in terms of model construction, one could also do an example similar to the one posted in #742 that differentiates with respect to a latent param that controls the yields, but with a more general set of models that explores the different types of systematics, though probably out of scope for this PR!

**Comment by phinate** [comment]
> thanks @phinate ccan you verify that this works too?
> 
> ```
>     def example_op(x):
>         y = pyhf.default_backend.power(x,2)
>         z = pyhf.tensorlib.sum(x)
>         return z
> ```

sure -- guessing you meant `z = pyhf.tensorlib.sum(y)`?

**Comment by phinate** [comment]
> Great tests.. Would be nice to also test that you can jit through this if the two backwnds are jax

nice idea! still works :)

```python3
def test_diffable_backend():
    pyhf.set_backend("jax", default=True)

    @jax.jit
    def example_op(x):
        y = pyhf.default_backend.astensor(x)
        return 2 * y

    assert jax.jacrev(example_op)([1.0]) == [2.0]

    @jax.jit
    def example_op2(x):
        y = pyhf.default_backend.power(x, 2)
        z = pyhf.tensorlib.sum(y)
        return z

    assert jax.jacrev(example_op2)(pyhf.tensorlib.astensor([2.0, 3.0])).tolist() == [
        4.0,
        6.0,
    ]


def test_diffable_backend_failure():
    pyhf.set_backend("numpy", default=True)
    pyhf.set_backend("jax")

    def example_op(x):
        y = pyhf.default_backend.astensor(x)
        return 2 * y

    with pytest.raises(Exception):
        jax.value_and_grad(example_op)(1.0)

    def example_op2(x):
        y = pyhf.default_backend.power(x, 2)
        z = pyhf.tensorlib.sum(y)
        return z

    with pytest.raises(Exception):
        jax.jacrev(example_op2)(pyhf.tensorlib.astensor([2.0, 3.0]))
```

**Comment by phinate** [comment]
> > @kratsg it seems like the first error encountered in the failure test has type `jax.errors.ConcretizationTypeError` instead of the one you have listed -- just what it hits first I guess from the numpy asarray call, doesn't even get to the tracer error. The other one works :)
> 
> That's confusing. It hits the other one first on my mac?

agh, that is confusing. could be os-dependent (yikes), since the CI and me are running on Ubuntu.

if you want, I don't think we lose anything by dropping the first failure mode of the test -- it's just to catch if we're mixing backends when diffing, which should always fail with any `default_backend` call if it's set to numpy, regardless of the op.

**Comment by phinate** [comment]
just for clarity, [here's the relevant page in the jax docs](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#non-array-inputs-numpy-vs-jax) that goes over this

**Comment by kratsg** [comment]
Huh, we do have `tests/test_simplemodels.py` but those don't cover the other backends.
